Add pdf-tools layer

Andre0991 opened the pull request:
I hope you guys like it.

jmiven commented on layers/+tools/pdf-tools/README.org:
```diff
@@ -0,0 +1,192 @@
+* Install
+
+** Prerequisites
+Linux is the only operational system officially supported, but it's possible to
```
"operating", not "operational" :smiley:

jmiven commented on layers/+tools/pdf-tools/README.org:
```diff
@@ -0,0 +1,192 @@
+  (setq-default dotspacemacs-configuration-layers '(pdf-tools))
+#+end_src
+
+After that, syncronise your configuration with ~SPC f e R~. This will pop up a
```
"synchronise"

robbyoconnor replied:
synchronize

robbyoconnor replied:
@Andre0991

Andre0991 replied:
Ops, missed that.

jmiven commented on layers/+tools/pdf-tools/README.org:
```diff
@@ -0,0 +1,192 @@
+pdf, displaying all headings.
+
+You can set annotations by selecting regions and right clicking on the buffer.
+Additionally, all annotation related functions are can be accessed with ~SPC a~ or
```
This is ~SPC m a~ :smiley:

jmiven commented on layers/+tools/pdf-tools/README.org:
```diff
@@ -0,0 +1,192 @@
+
+Tips:
+- You'll probably want to use =pdf-view-set-slice-from-bounding-box= (bounded to
+  ~SPC s b~ and ~M-m s b~.) This command trims the blank parts of the
```
~SPC m s b~

jmiven commented on layers/+tools/pdf-tools/README.org:
```diff
@@ -0,0 +1,192 @@
+Additionally, all annotation related functions are can be accessed with ~SPC a~ or
+~M-m a~.
+
+Searching with =occur= can be done by using =pdf-occur= (~SPC m o~ or ~M-m m o~).
```
I believe this is ~SPC m s o~

jmiven commented:
I sure like it, pdf-tools seems awesome compared to docview!

I was able to install the layer and run one or two commands. I'll try to test more in depth tomorrow.

Andre0991 commented:
Thanks @jmiven !

I fixed what you pointed out.

robbyoconnor commented:
:+1: :+1:

nixmaniack commented on layers/+tools/pdf-tools/README.org:
```diff
@@ -0,0 +1,192 @@
+
+Tips:
+- You'll probably want to use =pdf-view-set-slice-from-bounding-box= (bounded to
+  ~SPC m s b~ and ~M-m s b~.) This command trims the blank parts of the
```
s/bounded/bound/
s/M-m/M-m m/
I doubt you need to specify both evil and emacs bindings while mentioning them, evil should be enough.

robbyoconnor replied:
`M-m` is the leader key-chord in the emacs editing style so what @nixmaniack said

nixmaniack commented on layers/+tools/pdf-tools/extensions.el:
```diff
@@ -0,0 +1,31 @@
+;;; extensions.el --- pdf-tools Layer extensions File for Spacemacs
+;;
+;; Copyright (c) 2012-2014 Sylvain Benner
+;; Copyright (c) 2014-2015 Sylvain Benner & Contributors
```
s/2015/2016 :smile:

Andre0991 replied:
Fixed. Good catch :+1:

nixmaniack commented on layers/+tools/pdf-tools/packages.el:
```diff
@@ -0,0 +1,114 @@
+        "j"  'pdf-view-next-line-or-next-page
+        "k"  'pdf-view-previous-line-or-previous-page
+        "l"  'pdf-view-next-page-command
+        "h"  'pdf-view-previous-page-command
```
I would propose following here:
```
"j" 'pdf-view-next-line-or-next-page
"J" 'pdf-view-next-page
"k" 'pdf-view-previous-line-or-previous-page
"K" 'pdf-view-previous-page
"l" 'image-forward-hscroll
"h" 'image-backward-hscroll
```
So that we have access to horizontal scrolling in case needed.

robbyoconnor replied:
:+1:

Andre0991 replied:
In the current version, you'd get horizontal scrolling by using the arrows keys.

I tried to follow DocView (evilified) keybindings,  so `h` and `l` go to previous/next pages, which probably would be used more frequently than moving horizontally (but this is less consistent).

Let's wait and see if someone else suggests something...

DocView keybindings:
```
| ~h~         | previous page                            |
| ~H~         | adjust to height                         |
| ~j~         | next line                                |
| ~k~         | previous line                            |
| ~K~         | kill proc and buffer                     |
| ~l~         | next page                                |
| ~n~         | go to next search occurrence             |
| ~N~         | go to previous search occurrence         |
| ~P~         | fit page to window                       |
| ~r~         | revert                                   |
| ~W~         | adjust to width                          |
```

nixmaniack replied:
Woops! Arrow keys! I don't use them when being evil! :stuck_out_tongue_winking_eye: 

I use these bindings when I have to zoom in pdf and need to scroll around, having binding on homerow with normal hjkl navigation makes sense.

Andre0991 replied:
I removed the `M-m` references.

@nixmaniack I don't use them too.  The thing is that ideally the user would never need to horizontally scroll, since adjusting the zoom or fitting the page width would make it unnecessary.

@syl20bnr What do you think? Should we mimic DocView's keybindings? Perhaps change both? I set more things to be accessible through the leader key, which is different from what DocView currently does.

nixmaniack replied:
> The thing is that ideally the user would never need to horizontally scroll, since adjusting the zoom or fitting the page width would make it unnecessary.

Not in my case. I do zoom and need to move around. (Should be the case if anyone's viewing on small screen).

jmiven replied:
@nixmaniack :+1: 
Incidentally, these are also the keybindings used by the vi-like PDF reader zathura.

Andre0991 replied:
Right, I agree with `h` and `l`, it feels more consistent.

Now, for the upper case ones, I suggest this:
J -> Scroll page down
K -> Scroll page up
H -> Go to previous page
L -> Go to next page

H and L represent left and right, so it's like turning a real page.
I think J and K should scroll down/up because this is probably the most frequent action when reading a PDF. Normally we'd do this with `C-d` and `C-u` but this is cumbersome IMO.

Thanks for the suggestions.

robbyoconnor replied:
:+1:

Andre0991 replied:
One more thing: `o` is used for opening links in the Spacemacs conventions, but I'm using it for opening the outline mode here.

nixmaniack replied:
In that case, use `O`(capital `O`) for outline mode and `o` for opening links if that makes following conventions easier.

declanqian commented:
Awesome!!!

Andre0991 commented:
@declanqian Thanks :+1: 

Now `o` opens a link, which follows Spacemacs conventions. Opening the outline is now bound to `O`. I added a binding to night mode. I edited the README to follow some conventions.

robbyoconnor commented:
What's the status of this?
On 02/01/2016 06:09 PM, Andre0991 wrote:
> @declanqian <https://github.com/declanqian> Thanks :+1:
>
> Now |o| opens a link, which follows Spacemacs conventions. Opening the
> outline is now bound to |O|. I added a binding to night mode. I edited
> the README to follow some conventions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/4627#issuecomment-178248072>.
>

Andre0991 commented:
I created a transient state for pdf-tools.

I haven't pushed it yet because I actually didn't really put the commands on the code, I want to get some feedback from the interface first (and also because editing hydras is a pain, hehe).

Here's it:

![screenshot 2016-02-05 23 14 01](https://cloud.githubusercontent.com/assets/7083061/12863558/a6a69d58-cc5e-11e5-84fb-c00a9a7a7aee.png)

Any suggestions on changing words, positions, etc?

AlejandroCatalina commented:
It looks really nice man, well aligned ;). :+1:

robbyoconnor commented:
@syl20bnr status on this?!?!

Andre0991 commented:
> @syl20bnr status on this?!?!

:eyes: 

@nixmaniack I ended up using the keybindings you suggested for navigation. I also added `d` and `u` which scrolls like `C-d` and `C-u`.

I added the transient state to the code now.

A question: when I press `a` in the transient state, `which-key` pops up. How to stop it from interfering with the transient state?

nixmaniack commented:
> @nixmaniack I ended up using the keybindings you suggested for navigation. I also added d and u which scrolls like `C-d` and `C-u`.

Cool! :+1: 

> A question: when I press `a` in the transient state, `which-key` pops up. How to stop it from interfering with the transient state?

Ping @justbur for this.

justbur commented:
@Andre0991 You can set `which-key-inhibit` to `t` to prevent which-key from appearing, on entry to and exit from the transient state for example.

Andre0991 commented:
@justbr Thanks.

I tried this but it still shows up, am I missing something?

```
  (spacemacs/set-leader-keys-for-major-mode 'pdf-view-mode "." (lambda () (interactive) (let ((which-key-inhibit t)) (progn (message "which-key-inhibit value is %s" which-key-inhibit) (spacemacs/pdf-tools-manipulation-transient-state/body)))))
```

I put the message just for debugging purposes and it does show `which-key-inhibit value is t`.

justbur commented:
@Andre0991 I was thinking of this

```elisp
:on-enter (setq which-key-inhibit t)
:on-exit (setq which-key-inhibit nil)
```

Andre0991 commented:
@justbur Ah, that's better. It works now, thanks.

Andre0991 commented:
I added documentation for the transient state.

Well, I think it's kind of ready now. Feedback welcome.

Note: I'm using `evilified` to define the keybindings, so the ones that I don't remap just have the same keys as the original package.

izahn commented:
I'm really looking forward to this, I can finally get rid of my shabby `/private/pdf-tools` layer! I've been trying it out from @Andre0991's repo, and it is generally working well. One small niggle; searching with `/` doesn't allow me to navigate to next/previous matches. That is, pressing `/test<ret>` finds the first occurrence of `test`, but then pressing `n` does not take me to the next one.

Andre0991 commented:
@izahn Ah, yeah, I tried to use the way `doc-view` search is done Spacemacs but it didn't work for `pdf-tools`.

So I really don't know how to get `n` and `N` behave like it typically does. It is annotated as `TODO` in the code.

Let me know if something else doesn't work as expected.

myrjola commented:
@Andre0991, the transient state was an excellent addition. Looking forward to getting this merged. The layer works well for me.

chu- commented:
cool!

syl20bnr commented:
I will merge it soon :-)

paynito commented:
can't wait


--
Brandon Payne
http://about.me/brandonpayne




On Tue, Feb 23, 2016 at 2:55 PM, Sylvain Benner <notifications@github.com>
wrote:

> I will merge it soon :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/4627#issuecomment-187687361>.
>

syl20bnr commented:
Thank you :purple_heart:

At last I was able to merge it, I fought a bit poppler on the mac but in the end I succeeded in making it work.
I moved the key bindings `M-SPC` and `S-M-SPC` to the more conventional `SPC m .` and reorganized a bit the code.

Cherry-picked into develop branch, you can safely delete your branch.

nixmaniack commented:
Thanks @syl20bnr!

Andre0991 commented:
Thanks!

robbyoconnor commented:
Great work.. love it.

--Rob
Sent from my cell, please excuse any typos.
On Mar 4, 2016 10:56 AM, "André Peric Tavares" <notifications@github.com>
wrote:

> Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/4627#issuecomment-192332641>.
>

chu- commented:
Great! thanks.

robbyoconnor commented:
Hi,

I'd like to add you to my professional network on LinkedIn.

- Robert

Confirm that you know Robert: https://www.linkedin.com/e/v2?e=0-ini67o9j-t2&t=ssuw&tracking=eml-guest-invite-cta&ek=invite_guest&sharedKey=GrOgfRgm&invitationId=6130896981557596161








You received an invitation to connect. LinkedIn will use your email address to make suggestions to our members in features like People You May Know. Unsubscribe here: https://www.linkedin.com/e/v2?e=0-ini67o9j-t2&t=lun&midToken=AQFmrFVhGEd36Q&ek=invite_guest&loid=AQGKb6yv9_HvDAAAAVRVSbc4FuXT6YtIo5kevQPUlronNZHvpesr_SHCx30K7Kz2uX-sS6NH0EZXpl3aXLgV2HLMt2LA7Af3v-6lxzNrahcXyq2mLI-sMKwCbMlspc55XlHxl9TuAHK7OAmOl9I4eMJbRggwVKp-keuLzqj04HaQh5hqU4ZDhumewwWYuZQFVkYJ&eid=0-ini67o9j-t2


This email was sent to reply@reply.github.com.

If you need assistance or have questions, please contact LinkedIn Customer Service: https://www.linkedin.com/e/v2?e=0-ini67o9j-t2&a=customerServiceUrl&ek=invite_guest

&copy; 2016 LinkedIn Corporation, 2029 Stierlin Court, Mountain View CA 94043. LinkedIn and the LinkedIn logo are registered trademarks of LinkedIn.

robbyoconnor commented:
Hi,

I'd like to add you to my professional network on LinkedIn.

- Robert

Confirm that you know Robert: https://www.linkedin.com/e/v2?e=0-ini67ocu-e9&t=ssuw&tracking=eml-guest-invite-cta&ek=invite_guest&sharedKey=-G-UrA4k&invitationId=6130896963790524417








You received an invitation to connect. LinkedIn will use your email address to make suggestions to our members in features like People You May Know. Unsubscribe here: https://www.linkedin.com/e/v2?e=0-ini67ocu-e9&t=lun&midToken=AQFmrFVhGEd36Q&ek=invite_guest&loid=AQFzTuRqiIPsDgAAAVRVSb9kRERQI6qD01-GSnGHlirfglOzXU64ZbxWG1_gRCbNUzNSKjk6k7fRhkSwzTlU0tc15oAtdcUSt1ze95xDo1g0ry4NIM6-I_te6Q33l46QVQJb4djIGj5H_glKNM5UCBcaRxrlKmuR3XevuJv_aEWW85G4lR86VircTn4HTUmenZgI&eid=0-ini67ocu-e9


This email was sent to reply@reply.github.com.

If you need assistance or have questions, please contact LinkedIn Customer Service: https://www.linkedin.com/e/v2?e=0-ini67ocu-e9&a=customerServiceUrl&ek=invite_guest

&copy; 2016 LinkedIn Corporation, 2029 Stierlin Court, Mountain View CA 94043. LinkedIn and the LinkedIn logo are registered trademarks of LinkedIn.
